src/extractor/guidance/turn_analysis.cpp

```diff
@@ -95,6 +97,109 @@ inline bool isRampClass(EdgeID eid, const util::NodeBasedDynamicGraph &node_base
 
 } // namespace detail
 
+bool TurnAnalysis::isRotary(const NodeID nid) const
+{
+    // translate a node ID into its respective coordinate stored in the node_info_list
+    const auto getCoordinate = [this](const NodeID node)
+    {
+        return util::Coordinate(node_info_list[node].lon, node_info_list[node].lat);
+    };
+
+    unsigned roundabout_name = 0;
+    std::set<unsigned> connected_names;
+
+    const auto getNextOnRoundabout = [this, &roundabout_name, &connected_names](const NodeID node)
+    {
+        EdgeID continue_edge = SPECIAL_EDGEID;
+        for (const auto edge : node_based_graph.GetAdjacentEdgeRange(node))
+        {
+            const auto &edge_data = node_based_graph.GetEdgeData(edge);
+            if (!edge_data.reversed && edge_data.roundabout)
+            {
+                if (SPECIAL_EDGEID != continue_edge)
+                {
+                    // fork in roundabout
+                    return SPECIAL_EDGEID;
+                }
+                // roundabout does not keep its name
+                if (roundabout_name != 0 && roundabout_name != edge_data.name_id &&
+                    requiresNameAnnounced(name_table.GetNameForID(roundabout_name),
+                                          name_table.GetNameForID(edge_data.name_id)))
+                {
+                    return SPECIAL_EDGEID;
+                }
+
+                roundabout_name = edge_data.name_id;
+
+                continue_edge = edge;
+            }
+            else if (!edge_data.roundabout)
+            {
+                // remember all connected road names
+                connected_names.insert(edge_data.name_id);
+            }
+        }
+        return continue_edge;
+    };
+
+    // the roundabout radius has to be the same for all locations we look at it from
+    // to guarantee this, we search the full roundabout for its vertices
+    // and select the three smalles ids
+    std::vector<NodeID> roundabout_nodes;
+    roundabout_nodes.reserve(40);
+    roundabout_nodes.push_back(nid);
+    while (roundabout_nodes.size() < 40) // hard abort if something is wrong
```

Comment:
this value is chosen at random to avoid getting stuck in invalid constructs. I can think of no situation in which we would get 40 intersections in a roundabout, but if there is a strange loop within one, this hard-abort stops it from happening.